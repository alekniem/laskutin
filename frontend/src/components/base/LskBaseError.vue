<template>
    <div class="red--text font-weight-bold">
        <slot />
    </div>
</template>

<script>
export default {
    name: "LskBaseError",
};
</script>
