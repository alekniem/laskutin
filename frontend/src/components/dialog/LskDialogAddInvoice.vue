<template>
    <v-dialog max-width="700" persistent v-model="dialogVisible">
        <ValidationObserver #default="{ handleSubmit, invalid }" ref="observer">
            <v-form>
                <v-card>
                    <v-card-title :id="cid + 'Title'">
                        {{ title }}
                    </v-card-title>
                    <v-card-text>
                        <v-row dense>
                            <v-col>
                                <LskInputDate
                                    :cid="cid + 'InvoiceDateInput'"
                                    label="Invoice date"
                                    vid="invoice_date"
                                    v-model="inputData.invoice_date"
                                />
                            </v-col>
                            <v-col>
                                <LskInputDate
                                    :cid="cid + 'DueDateInput'"
                                    label="Due date"
                                    vid="due_date"
                                    v-model="inputData.due_date"
                                />
                            </v-col>
                        </v-row>
                        <v-row dense>
                            <v-col>
                                <LskInputSelectCustomer
                                    :cid="cid + 'CustomerIdInput'"
                                    label="Customer"
                                    vid="customer_id"
                                    v-model="inputData.customer_id"
                                />
                            </v-col>
                        </v-row>
                        <v-row dense>
                            <v-col>
                                <LskInputSelectBiller
                                    :cid="cid + 'BillerIdInput'"
                                    label="Biller"
                                    vid="biller_id"
                                    v-model="inputData.biller_id"
                                />
                            </v-col>
                        </v-row>
                        <template v-if="otherErrors">
                            <LskBaseError
                                v-for="(error, key) in otherErrors"
                                :key="key"
                            >
                                {{ error | trans }}
                            </LskBaseError>
                        </template>
                    </v-card-text>
                    <v-card-actions>
                        <v-spacer />
                        <LskInputButton
                            @click.native="processCancel"
                            :cid="cid + 'CancelButton'"
                            buttonText="Cancel"
                        />
                        <LskInputButton
                            @click.native="handleSubmit(processSave)"
                            :cid="cid + 'SaveButton'"
                            :disabled="invalid || saveButtonDisabled"
                            buttonText="Save"
                        />
                    </v-card-actions>
                </v-card>
            </v-form>
        </ValidationObserver>
    </v-dialog>
</template>

<script>
import LskDialogAddOrEdit from "@/mixins/LskDialogAddOrEdit";

export default {
    name: "LskDialogAddInvoice",

    mixins: [LskDialogAddOrEdit],

    props: {
        cid: { type: String, required: true },
    },

    data() {
        return {
            fields: ["invoice_date", "due_date", "biller_id", "customer_id"],
        };
    },

    computed: {
        apiActionName() {
            return "invoiceStore";
        },

        title() {
            return this.trans("Add new invoice");
        },
    },
};
</script>
