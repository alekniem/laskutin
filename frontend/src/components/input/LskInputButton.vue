<template>
    <v-btn :color="color" :disabled="disabled" :id="cid" depressed small>
        {{ buttonText | trans }}
    </v-btn>
</template>

<script>
export default {
    name: "LskInputButton",

    props: {
        buttonText: { type: String, required: true },
        cid: { type: String, required: true },
        color: { type: String, required: false, default: "primary" },
        disabled: { type: Boolean, required: false, default: false },
    },
};
</script>
