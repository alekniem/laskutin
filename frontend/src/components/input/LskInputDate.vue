<template>
    <ValidationProvider
        :customMessages="{ regex: trans('Invalid date') }"
        :rules="{ regex: /^\d{1,2}\.\d{1,2}\.\d{4}$/ }"
        :vid="vid"
        ref="provider"
        tag="div"
        #default="{ errors, valid }"
    >
        <v-text-field
            :error-messages="errors"
            :id="cid"
            :label="trans(label)"
            :success="valid"
            v-model="textFieldValue"
        />
    </ValidationProvider>
</template>

<script>
import { ValidationProvider } from "vee-validate";

export default {
    name: "LskInputDate",

    components: {
        ValidationProvider,
    },

    props: {
        cid: { type: String, required: true },
        label: { type: String, required: true },
        value: { type: String, required: true },
        vid: { type: String, required: true },
    },

    data() {
        return {
            textFieldValue: null,
        };
    },

    watch: {
        textFieldValue() {
            if (/^\d{1,2}\.\d{1,2}\.\d{4}$/.test(this.textFieldValue)) {
                let parts = this.textFieldValue.split(".");
                let date = new Date(parts[2], parts[1] - 1, parts[0]);
                let result = this.formatDate(date, "YYYY-MM-DD");
                this.$emit("input", result);
            }
        },
    },

    mounted() {
        if (/^\d{4}-\d{2}-\d{2}$/.test(this.value)) {
            this.textFieldValue = this.formatDate(this.value, "D.M.YYYY");
        } else {
            this.textFieldValue = this.value;
        }
    },
};
</script>
