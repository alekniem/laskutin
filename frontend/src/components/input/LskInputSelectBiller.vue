<template>
    <ValidationProvider
        :customMessages="{ integer: trans('Invalid biller id') }"
        :vid="vid"
        ref="provider"
        rules="integer"
        tag="div"
        #default="{ errors, valid }"
    >
        <v-select
            @change="$emit('input', $event)"
            :error-messages="errors"
            :id="cid"
            :items="items"
            :label="trans(label)"
            :success="valid"
            :value="value"
        />
    </ValidationProvider>
</template>

<script>
import lsk_api from "@/plugins/lsk_api";
import { ValidationProvider } from "vee-validate";

export default {
    name: "LskInputSelectBiller",

    components: {
        ValidationProvider,
    },

    props: {
        cid: { type: String, required: true },
        label: { type: String, required: true },
        value: { type: [String, Number], required: true },
        vid: { type: String, required: true },
    },

    data() {
        return {
            items: [],
        };
    },

    mounted() {
        lsk_api
            .query("billerIndex", null, null)
            .then((response) => this.parseResponse(response.data))
            .catch(() => (this.items = []));
    },

    methods: {
        parseResponse(data) {
            this.items = data.map((line) => ({
                value: line.id,
                text:
                    line.id +
                    ". " +
                    line.name +
                    ", " +
                    line.address_line_one +
                    ", " +
                    line.address_line_two,
            }));
        },
    },
};
</script>
