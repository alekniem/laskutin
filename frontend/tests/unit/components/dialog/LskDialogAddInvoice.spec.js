import axios from "axios";
import flushPromises from "flush-promises";
import LskBaseError from "@/components/base/LskBaseError";
import LskDialogAddInvoice from "@/components/dialog/LskDialogAddInvoice";
import LskInputButton from "@/components/input/LskInputButton";
import LskInputDate from "@/components/input/LskInputDate";
import LskInputSelectBiller from "@/components/input/LskInputSelectBiller";
import LskInputSelectCustomer from "@/components/input/LskInputSelectCustomer";
import { createWrapper } from "UT/lsk_test_helper";

jest.mock("axios");

axios.get.mockImplementation(() => Promise.resolve());

const stubs = {
    LskBaseError,
    LskInputButton,
    LskInputDate,
    LskInputSelectBiller,
    LskInputSelectCustomer,
};

describe("LskDialogAddInvoice", () => {
    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogAddInvoice, {
            propsData: {
                cid: "TEST_CID",
                showDialog: false,
            },
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog not visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogAddInvoice, {
            propsData: {
                cid: "TEST_CID",
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogAddInvoice, {
            propsData: {
                cid: "TEST_CID",
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();

        const invoiceDateInput = wrapper.find("#TEST_CIDInvoiceDateInput");
        invoiceDateInput.element.value = "INVALID INVOICE DATE";
        invoiceDateInput.trigger("input");
        const dueDateInput = wrapper.find("#TEST_CIDDueDateInput");
        dueDateInput.element.value = "INVALID DUE DATE";
        dueDateInput.trigger("input");
        await flushPromises();

        wrapper.vm._data.inputData.customer_id = "INVALID CUSTOMER ID";
        wrapper.vm._data.inputData.biller_id = "INVALID BILLER ID";
        wrapper.vm._data.otherErrors = ["ERROR ONE", "ERROR TWO"];
        await flushPromises();

        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("error messages");
    });

    it("should have correct title", async () => {
        const wrapper = createWrapper(LskDialogAddInvoice, {
            propsData: {
                cid: "TEST_CID",
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        expect(wrapper.find(".v-card__title").text()).toBe("Add new invoice");
    });
});
