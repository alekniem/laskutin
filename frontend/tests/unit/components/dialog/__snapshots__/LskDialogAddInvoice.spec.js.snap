// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LskDialogAddInvoice should match snapshot: dialog not visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
</div>
`;

exports[`LskDialogAddInvoice should match snapshot: dialog visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                Add new invoice
              </div>
              <div class="v-card__text">
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted success--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDInvoiceDateInput"
                                     class="v-label theme--light success--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Invoice date
                              </label>
                              <input id="TEST_CIDInvoiceDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light success--text">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted success--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDDueDateInput"
                                     class="v-label theme--light success--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Due date
                              </label>
                              <input id="TEST_CIDDueDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light success--text">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted v-select success--text">
                        <div class="v-input__control">
                          <div role="button"
                               aria-haspopup="listbox"
                               aria-expanded="false"
                               aria-owns="list-24"
                               class="v-input__slot"
                          >
                            <div class="v-select__slot">
                              <label for="TEST_CIDCustomerIdInput"
                                     class="v-label theme--light success--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Customer
                              </label>
                              <div class="v-select__selections">
                                <input id="TEST_CIDCustomerIdInput"
                                       readonly="readonly"
                                       type="text"
                                       aria-readonly="false"
                                       autocomplete="off"
                                >
                              </div>
                              <div class="v-input__append-inner">
                                <div class="v-input__icon v-input__icon--append">
                                  <i aria-hidden="true"
                                     class="v-icon notranslate mdi mdi-menu-down theme--light success--text"
                                  >
                                  </i>
                                </div>
                              </div>
                              <input type="hidden">
                            </div>
                            <div class="v-menu">
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light success--text">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted v-select success--text">
                        <div class="v-input__control">
                          <div role="button"
                               aria-haspopup="listbox"
                               aria-expanded="false"
                               aria-owns="list-31"
                               class="v-input__slot"
                          >
                            <div class="v-select__slot">
                              <label for="TEST_CIDBillerIdInput"
                                     class="v-label theme--light success--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Biller
                              </label>
                              <div class="v-select__selections">
                                <input id="TEST_CIDBillerIdInput"
                                       readonly="readonly"
                                       type="text"
                                       aria-readonly="false"
                                       autocomplete="off"
                                >
                              </div>
                              <div class="v-input__append-inner">
                                <div class="v-input__icon v-input__icon--append">
                                  <i aria-hidden="true"
                                     class="v-icon notranslate mdi mdi-menu-down theme--light success--text"
                                  >
                                  </i>
                                </div>
                              </div>
                              <input type="hidden">
                            </div>
                            <div class="v-menu">
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light success--text">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        disabled="disabled"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDSaveButton"
                >
                  <span class="v-btn__content">
                    Save
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`LskDialogAddInvoice should match snapshot: error messages 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                Add new invoice
              </div>
              <div class="v-card__text">
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDInvoiceDateInput"
                                     class="v-label v-label--active theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Invoice date
                              </label>
                              <input id="TEST_CIDInvoiceDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Invalid date
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDDueDateInput"
                                     class="v-label v-label--active theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Due date
                              </label>
                              <input id="TEST_CIDDueDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Invalid date
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted v-select error--text">
                        <div class="v-input__control">
                          <div role="button"
                               aria-haspopup="listbox"
                               aria-expanded="false"
                               aria-owns="list-61"
                               class="v-input__slot"
                          >
                            <div class="v-select__slot">
                              <label for="TEST_CIDCustomerIdInput"
                                     class="v-label theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Customer
                              </label>
                              <div class="v-select__selections">
                                <input id="TEST_CIDCustomerIdInput"
                                       readonly="readonly"
                                       type="text"
                                       aria-readonly="false"
                                       autocomplete="off"
                                >
                              </div>
                              <div class="v-input__append-inner">
                                <div class="v-input__icon v-input__icon--append">
                                  <i aria-hidden="true"
                                     class="v-icon notranslate mdi mdi-menu-down theme--light error--text"
                                  >
                                  </i>
                                </div>
                              </div>
                              <input type="hidden"
                                     value="INVALID CUSTOMER ID"
                              >
                            </div>
                            <div class="v-menu">
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Invalid customer id
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted v-select error--text">
                        <div class="v-input__control">
                          <div role="button"
                               aria-haspopup="listbox"
                               aria-expanded="false"
                               aria-owns="list-68"
                               class="v-input__slot"
                          >
                            <div class="v-select__slot">
                              <label for="TEST_CIDBillerIdInput"
                                     class="v-label theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Biller
                              </label>
                              <div class="v-select__selections">
                                <input id="TEST_CIDBillerIdInput"
                                       readonly="readonly"
                                       type="text"
                                       aria-readonly="false"
                                       autocomplete="off"
                                >
                              </div>
                              <div class="v-input__append-inner">
                                <div class="v-input__icon v-input__icon--append">
                                  <i aria-hidden="true"
                                     class="v-icon notranslate mdi mdi-menu-down theme--light error--text"
                                  >
                                  </i>
                                </div>
                              </div>
                              <input type="hidden"
                                     value="INVALID BILLER ID"
                              >
                            </div>
                            <div class="v-menu">
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Invalid biller id
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="red--text font-weight-bold">
                  ERROR ONE
                </div>
                <div class="red--text font-weight-bold">
                  ERROR TWO
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDSaveButton"
                        disabled="disabled"
                >
                  <span class="v-btn__content">
                    Save
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;
